File: PythonSolutions/Neetcode150/graph_problems/207_course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        preMap = defaultdict(list)

        # Create the adjacent list graph where course is pointing to pre-req
            # This tells us: "To find what courses we need to complete for this course"
            # If we flip it (pre -> course): "What courses will this unlock?"
        for crs, pre in prerequisites:
            preMap[crs].append(pre)

        visit = set() # Current path
        done = set()  # Nodes that have been fully visited

        def dfs(crs):
            # Base case 1: Cycle detected
            if crs in visit:
                return False
            # Base case 2: Already visited in another cycle
            if crs in done:
                return True
            
            # Visit the course node and add it to current path
            visit.add(crs)

            # Go through all pre-requisites for the course and see if it can be completed
            for pre in preMap[crs]:
                # Check if a the pre-requisite is found in a visit making it impossible to finish the course
                if not dfs(pre):
                    return False
            
            # Remove course from visit after we have visited all nodes for this course
            visit.remove(crs)
            # Mark it as done since we have visited all pre-requisite nodes for this course
            done.add(crs)

            # Course was able to be completed, mark as true
            return True

        # Go through each node in the schedule
        for crs in range(numCourses):
            # If a cycle was detected then we just couldn't finish the course plan
            if not dfs(crs):
                return False
        
        return True

File: TypescriptSolutions/Neetcode150/graph_grid_problems/207_course-schedule.ts
export {};
const canFinish = (numCourses: number, prerequisites: number[][]): boolean => {
    const preMap: Record<number, number[]> = {};

    for (let i = 0; i < numCourses; i++) preMap[i] = [];

    // Course -> Prereq
    for (const [crs, pre] of prerequisites) {
        preMap[crs].push(pre);
    }

    const visit = new Set<number>();
    const completed = new Set<number>();

    const dfs = (crs: number): boolean => {
        if (visit.has(crs)) return false;
        if (completed.has(crs)) return true;

        visit.add(crs);

        for (const pre of preMap[crs]) {
            if (!dfs(pre)) return false;
        }

        completed.add(crs);
        visit.delete(crs);

        return true;
    };

    for (let crs = 0; crs < numCourses; crs++) {
        if (!dfs(crs)) return false;
    }

    return true;
};